File: 作业3-合并两个有序链表-结果仍然有序/java/Solution.java
class ListNode {
     int val;
     ListNode next;
     ListNode(int x) { val = x; }
}

public class Solution {
	public void main(int[] args) {

	}

	/**
	 * https://leetcode-cn.com/problems/merge-two-sorted-lists/
	 * @param  list1 [description]
	 * @param  list2 [description]
	 * @return       [description]
	 */
	public ListNode mergeTwoListIteration(ListNode list1, ListNode list2) {
		ListNode preHead = new ListNode(-1);

		ListNode prev = preHead;
		while (list1 != null && list2 != null) {
			if (list1.val < list2.val) {
				prev.next = list1;
				list1 = list1.next;
			} else {
				prev.next = list2;
				list2 = list2.next;
			}
			prev = prev.next;
		}

		preHead.next = list1 == null ? list2 : list1;
		return preHead.next;
	}

	public ListNode mergeTwoListRecursion(ListNode list1, ListNode list2) {
		if (list1 == null) {
			return list2;
		} else if (list2 == null) {
			return list1;
		} else (list1.val  < list2.val) {
			list1.next = mergeTwoListRecursion(list1.next, list2);
			return list1;
		} else {
			list2.next = mergeTwoListRecursion(list1, list2.next);
			return list2;
		}
	}
}

File: 作业3-合并两个有序链表-结果仍然有序/php/Solution.php
<?php

class Solution {
	function merge($l1, $l2) {
		$preHead = new ListNode(null);
		$cur = $preHead;
		while ($l1 != null && $l2 != null) {
			if ($l1->val <= $l2->val) {
				$cur->next = $l1;
				$l1 = $l1->next;
			} else {
				$cur->next = $l2;
				$l2 = $l2->next;
			}
			$cur = $cur->next;
		}

		$cur->next = $l1 == null ? $l2 : $l1;
		return $preHead->next;
	}

	function mergeTwoLists($l1, $l2) {
	    // 递归解法
	    // 递归函数的含义：返回当前两个链表合并之后的头节点(每一层都返回排序好的链表头)
	    if ($l1 === null) {
	        return $l2;
	    }

	    if ($l2 === null) {
	        return $l1;
	    }

	    if ($l1->val < $l2->val) {
	        $l1->next = $this->mergeTwoLists($l1->next, $l2);
	        return $l1;
	    } else {
	        $l2->next = $this->mergeTwoLists($l1, $l2->next);
	        return $l2;
	    }
	}
}